### PR TITLE
Add a disk cache for transpiled files

### PR DESCRIPTION
**Pull request by dumbmatter**
I tried to address all the ideas in #5 by caching transpiled JS files in node_modules/.cache/import-jsx. babel-loader [has a cache that does basically the same thing](https://github.com/babel/babel-loader/blob/f63f4d87845d5baaeb67107646fd5ae0e5aae2cd/src/cache.js), I tried to make the minimal changes needed to work here (like getting rid of all the async stuff).

Some rough benchmarks...

| File Size | First run | Cached |
| - | - | - |
| 0.1kb | 0.4s | 0.07s |
| 30kb | 1.4s | 0.09s |
| 300kb | 24s | 0.14s |

That's a pretty nice gain, even for tiny files!

I know this PR hurts your test coverage... it's just for all the edge cases in cache.js, like if you don't have write permission to node_modules. Not sure if it's worth doing anything about that.

(Also thanks for Ink, I tried it for the first time yesterday and it kinda blew my mind how easy it was to make some pretty CLI output! But the half-second delay added when using import-jsx was mildly annoying, and I didn't want to add a build step, so here I am..)

### CODE REVIEW

**Comment by dumbmatter** [comment]
Also, looks like a dependency of find-cache-dir does not support Node 6, so we'd either need to work around that or bump up the minimum supported version of Node for this package.

**Comment by coreyfarrell** [comment]
The current version of find-cache-dir requires node.js 8 (I think the author officially only supports LTS releases so 8.9.0+).  find-cache-dir 2.x supports node.js 6.

File: cache.js [hunk 1]
```diff
@@ -0,0 +1,103 @@
+/**
```
Comment (vadimdemedes) [inline cache.js]:
```suggestion
'use strict';
/**
```

File: cache.js [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+ */
+const fs = require('fs');
+const os = require('os');
+const path = require('path');
+const crypto = require('crypto');
+const mkdirp = require('mkdirp');
+const findCacheDir = require('find-cache-dir');
+const transform = require('./transform');
+
+let directory = null;
```
Comment (vadimdemedes) [inline cache.js]:
```suggestion
let directory;
```

File: cache.js [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+ * Build the filename for the cached file
+ *
+ * @param  {String} source  Original contents of the file to be cached
+ * @param  {Object} options Options passed to importJsx
+ * @param  {String} version Version of import-jsx
+ *
+ * @return {String}
+ */
+const filename = (source, options, version) => {
+	const hash = crypto.createHash('md4');
```
Comment (vadimdemedes) [inline cache.js]:
```suggestion
	const hash = crypto.createHash('md4');
	const contents = JSON.stringify({source, options, version});
 	hash.update(contents);
 	
 	return hash.digest('hex') + '.js';
```

File: cache.js [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+		version
+	} = params;
+
+	const file = path.join(directory, filename(source, options, version));
+
+	try {
+		// No errors mean that the file was previously cached
+		// we just need to return it
+		return fs.readFileSync(file).toString();
+	} catch (err) {}
```
Comment (vadimdemedes) [inline cache.js]:
```suggestion
	} catch (error) {}
```

File: cache.js [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+
+	const fallback = directory !== os.tmpdir();
+
+	// Make sure the directory exists.
+	try {
+		mkdirp.sync(directory);
+	} catch (err) {
```
Comment (vadimdemedes) [inline cache.js]:
```suggestion
	} catch (error) {
```

File: cache.js [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+		throw err;
+	}
+
+	// Otherwise just transform the file
+	// return it to the user asap and write it in cache
+	const result = transform(source, options, modulePath);
+
+	try {
+		fs.writeFileSync(file, result);
+	} catch (err) {
```
Comment (vadimdemedes) [inline cache.js]:
```suggestion
	} catch (error) {
```

File: cache.js [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+ * Retrieve file from cache, or create a new one for future reads
+ *
+ * @param  {Object}   params
+ * @param  {String}   params.modulePath
+ * @param  {String}   params.source     Original contents of the file to be cached
+ * @param  {Object}   params.options    Options passed to importJsx
+ * @param  {String}   params.version    Version of import-jsx
+ */
+module.exports = params => {
+	if (directory === null) {
```
Comment (vadimdemedes) [inline cache.js]:
```suggestion
	if (!directory) {
```

File: transform.js [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+let jsxTransform;
+
+const transform = (source, options, modulePath) => {
+	if (!babel) {
+		babel = require('@babel/core');
+		destructuringTransform = require('@babel/plugin-transform-destructuring');
+		restSpreadTransform = require('@babel/plugin-proposal-object-rest-spread');
+		jsxTransform = require('@babel/plugin-transform-react-jsx');
+	}
+	if (source.includes('React')) {
```
Comment (vadimdemedes) [inline transform.js]: Insert an empty line above.

**Comment by vadimdemedes** [comment]
Could you also add a few simple tests to verify caching works as expected?

**Comment by vadimdemedes** [comment]
> Also, looks like a dependency of find-cache-dir does not support Node 6, so we'd either need to work around that or bump up the minimum supported version of Node for this package.

Don't worry about this, I'll remove support for v6.x after merging this.

**Comment by dumbmatter** [comment]
Thanks for the feedback! I think I made all your requested changes.

Renaming `err` to `error` caused lint errors - do you want to use `err` or add an override to the lint settings?

**Comment by vadimdemedes** [comment]
> Renaming err to error caused lint errors - do you want to use err or add an override to the lint settings?

Hm, didn't expect that, I'll rename back.

File: test/test.js [hunk 9]
```diff
@@ -74,8 +81,26 @@ test('works when destructuring isnt available natively', t => {
...
-test('parse React fragments', t => {
-	t.notThrows(() => {
-		importJsx(fixturePath('react-fragment'));
-	});
+const diskCacheFile = `${diskCacheDirectory}/f478c8d5f1d242d6e659f8c40f33374c.js`;
+
+test('creates appropriate cache file on disk', t => {
+	t.false(fs.existsSync(diskCacheFile));
+
+	importJsx(fixturePath('for-cache'), {cache: false});
```
Comment (vadimdemedes) [inline test/test.js]: If cache is turned off, why does it create a cache file?

File: test/test.js [hunk 10]
```diff
@@ -74,8 +81,26 @@ test('works when destructuring isnt available natively', t => {
...
+
+	t.true(fs.existsSync(diskCacheFile));
+});
+
+test('uses cache file from disk', t => {
+	const text = importJsx(fixturePath('for-cache'), {cache: false});
```
Comment (vadimdemedes) [inline test/test.js]: Same here, cache is disabled, yet it's still working.

File: test/test.js (hunk 9 above)
Comment (dumbmatter) [reply test/test.js]: The "cache" option is for the existing in memory cache, which this PR doesn't change. I didn't add an option for the new disk cache. If you think it's a good idea, I could rename the existing "cache" option to "cacheMemory" and add a new option "cacheDisk" to enable the new disk cache.

Comment (vadimdemedes) [reply test/test.js]: I think `cache` option should enable/disable both disk and transform cache.

Comment (dumbmatter) [reply test/test.js]: That makes sense. I just made that change - now the `cache` option affects both caches.

**Comment by vadimdemedes** [comment]
Lovely, thank you! Great work!
